R/shinyapp.R

```diff
@@ -290,33 +308,77 @@ initAutoReloadMonitor <- function(dir) {
     return(function(){})
   }
 
-  filePattern <- getOption("shiny.autoreload.pattern",
-    ".*\\.(r|html?|js|css|png|jpe?g|gif)$")
+  filePattern <- getOption(
+    "shiny.autoreload.pattern",
+    ".*\\.(r|html?|js|css|png|jpe?g|gif)$"
+  )
 
-  lastValue <- NULL
-  observeLabel <- paste0("File Auto-Reload - '", basename(dir), "'")
-  obs <- observe(label = observeLabel, {
-    files <- sort_c(
-      list.files(dir, pattern = filePattern, recursive = TRUE, ignore.case = TRUE)
-    )
-    times <- file.info(files)$mtime
-    names(times) <- files
-
-    if (is.null(lastValue)) {
-      # First run
-      lastValue <<- times
-    } else if (!identical(lastValue, times)) {
-      # We've changed!
-      lastValue <<- times
+  
+  if (is_installed("watcher")) {
+    check_for_update <- function(paths) {
+      paths <- grep(
+        filePattern,
+        paths,
+        ignore.case = TRUE,
+        value = TRUE
+      )
+      
+      if (length(paths) == 0) {
+        return()
+      }
+      
+      cachedAutoReloadLastChanged$set()
       autoReloadCallbacks$invoke()
     }
+    
+    # [garrick, 2025-02-20] Shiny <= v1.10.0 used `invalidateLater()` with an
+    # autoreload.interval in ms. {watcher} instead uses a latency parameter in
+    # seconds, which serves a similar purpose and that I'm keeping for backcompat.
+    latency <- getOption("shiny.autoreload.interval", 250) / 1000
+    watcher <- watcher::watcher(dir, check_for_update, latency = latency)
+    watcher$start()
+    onStop(watcher$stop)
+  } else {
+    # Fall back to legacy observer behavior
+    if (!isFALSE(getOption("shiny.autoreload.legacy_warning", TRUE))) {
+      cli::cli_warn(
+        c(
+          "Using legacy autoreload file watching. Please install {.pkg watcher} for a more performant autoreload file watcher.",
+          "i" = "Set {.run options(shiny.autoreload.legacy_warning = FALSE)} to suppress this warning."
+        ),
+        .frequency = "regularly",
+        .frequency_id = "shiny.autoreload.legacy_warning"
+      )
+    }
```

Comment:
@cpsievert Should this use `get_devmode_option()`? And if so, what would that look like?

Reply:
btw, I _don't think_ `shiny.autoreload.legacy_warning` should be documented anywhere, but I do envision someone wanting to suppress the message entirely if they don't want to use watcher.

Reply:
It doesn't seem like `get_devmode_option()` is necessary. The only reason to use it is if you want the option's default to be different when `devmode(TRUE)` is activated